sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/gradle/FileIndexMap.kt

```diff
@@ -0,0 +1,72 @@
+package com.squareup.sqldelight.intellij.gradle
+
+import com.intellij.openapi.externalSystem.util.ExternalSystemApiUtil
+import com.intellij.openapi.module.Module
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Task
+import com.squareup.sqldelight.core.SqlDelightFileIndex
+import com.squareup.sqldelight.core.SqlDelightProjectService
+import com.squareup.sqldelight.core.dialectPreset
+import com.squareup.sqldelight.intellij.FileIndex
+import com.squareup.sqldelight.intellij.SqlDelightFileIndexImpl
+import org.jetbrains.plugins.gradle.service.execution.GradleExecutionHelper
+import org.jetbrains.plugins.gradle.settings.DistributionType
+import org.jetbrains.plugins.gradle.settings.GradleExecutionSettings
+import timber.log.Timber
+
+internal class FileIndexMap {
+  private val fileIndices = mutableMapOf<String, SqlDelightFileIndex>()
+
+  private var initializing = false
+
+  operator fun get(module: Module): SqlDelightFileIndex {
+    val projectPath = ExternalSystemApiUtil.getExternalProjectPath(module) ?: return defaultIndex
+    val result = fileIndices[projectPath]
+    if (result != null) return result
+    synchronized(this) {
+      if (!initializing) {
+        initializing = true
+        ProgressManager.getInstance().run(FetchModuleModels(module, projectPath))
+      }
+    }
+    return fileIndices[projectPath] ?: defaultIndex
+  }
+
+  private inner class FetchModuleModels(
+    private val module: Module,
+    private val projectPath: String
+  ) : Task.Backgroundable(
+      /* project = */ module.project,
+      /* title = */ "Importing ${module.name} SQLDelight"
+  ) {
+    override fun run(indicator: ProgressIndicator) {
+      val executionSettings = GradleExecutionSettings(
+          /* gradleHome = */ null,
+          /* serviceDirectory = */ null,
+          /* distributionType = */ DistributionType.DEFAULT_WRAPPED,
+          /* isOfflineWork = */ false
+      )
+      fileIndices.putAll(GradleExecutionHelper().execute(projectPath, executionSettings) { connection ->
```

Comment:
IntelliJ openapi exposes this API for running off a GradleConnection, so we don't need to bundle it on our classpath and we can share the same daemon pool as IntelliJ